File: UIPeriodicTable/CPlusPlus/Resources/Components/code_uiwidget.cpp
// UIWidget application source code
#include "PeriodicApp.h"

void PeriodicApp::setup_uiwidget( UIWidget *layout )
{
    PODVector<UIWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for (unsigned ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, ATOMIC_HANDLER(PeriodicApp, HandleUiwidgetEvent ));
}

void PeriodicApp::HandleUiwidgetEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    UIWidget* widget = static_cast<UIWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;
    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uiwidgetcode" )
        {
            AppLog( "UIWidget support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Components/code_uiwidget.cpp", widget->GetParent() );
        }
        if (widget->GetId() ==  "uiwidgetlayout" )
        {
            AppLog( "UIWidget support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_uiwidget.ui.txt", widget->GetParent() );
        }
    }
}

File: UIPeriodicTable/CSharp/Resources/Components/code_uiwidget.cs
// UIWidgetapplication source code
using System;
using AtomicEngine;

public class code_uiwidget : CSComponent {

    public void Setup( UIWidget  layout )
    {
        var dest = new AtomicEngine.Vector<AtomicEngine.UIWidget>();
        layout.SearchWidgetClass( "TBButton", dest );
        for (var ii = 0; ii < dest.Size; ii++) {
            dest[ii].SubscribeToEvent<WidgetEvent> (dest [ii], HandleUiwidgetEvent );
        }
    }

    private static void HandleUiwidgetEvent ( WidgetEvent ev )
    {
        UIWidget widget = (UIWidget)ev.Target;
        if ( widget.Equals(null)) return;
        if ( ev.Type == UI_EVENT_TYPE.UI_EVENT_TYPE_CLICK) {
            if (widget.GetId() == "uiwidgetcode" ) {
                AtomicMain.AppLog( "UIWidgetsupport : " + widget.GetId() + " was pressed " );
                AtomicMain.ViewCode ( "Components/code_uiwidget.cs", widget.GetParent() );
            }
            if (widget.GetId() ==  "uiwidgetlayout" ) {
                AtomicMain.AppLog( "UIWidgetsupport : " + widget.GetId() + " was pressed ");
                AtomicMain.ViewCode ( "Scenes/layout_uiwidget.ui.txt", widget.GetParent() );
            }
        }
    }
}

